src/aggregator/aggregator/handler/writer/protobuf.go

```diff
@@ -0,0 +1,185 @@
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+
+package writer
+
+import (
+	"errors"
+	"math/rand"
+
+	"github.com/m3db/m3/src/metrics/encoding/protobuf"
+	"github.com/m3db/m3/src/metrics/metric/aggregated"
+	"github.com/m3db/m3/src/msg/producer"
+	"github.com/m3db/m3x/clock"
+	murmur3 "github.com/m3db/stackmurmur3"
+
+	"github.com/uber-go/tally"
+)
+
+var (
+	errWriterClosed = errors.New("writer is closed")
+)
+
+type protobufWriterMetrics struct {
+	writerClosed  tally.Counter
+	encodeSuccess tally.Counter
+	encodeErrors  tally.Counter
+	routeSuccess  tally.Counter
+	routeErrors   tally.Counter
+}
+
+func newProtobufWriterMetrics(scope tally.Scope) protobufWriterMetrics {
+	encodeScope := scope.SubScope("encode")
+	routeScope := scope.SubScope("route")
+	return protobufWriterMetrics{
+		writerClosed:  scope.Counter("writer-closed"),
+		encodeSuccess: encodeScope.Counter("success"),
+		encodeErrors:  encodeScope.Counter("errors"),
+		routeSuccess:  routeScope.Counter("success"),
+		routeErrors:   routeScope.Counter("errors"),
+	}
+}
+
+type idShardFn func([]byte) uint32
+
+// protobufWriter encodes data and routes them to the backend.
+// protobufWriter is not thread safe.
+type protobufWriter struct {
+	encodingTimeSamplingRate float64
+	encoder                  protobuf.AggregatedEncoder
+	p                        producer.Producer
+	numShards                uint32
+
+	closed  bool
+	m       aggregated.MetricWithStoragePolicy
+	rand    *rand.Rand
+	metrics protobufWriterMetrics
+
+	nowFn   clock.NowFn
+	randFn  randFn
+	shardFn idShardFn
+}
+
+// NewProtobufWriter creates a writer that encodes metric in protobuf.
+func NewProtobufWriter(
+	producer producer.Producer,
+	opts Options,
+) (Writer, error) {
+	nowFn := opts.ClockOptions().NowFn()
+	instrumentOpts := opts.InstrumentOptions()
+	w := &protobufWriter{
+		encodingTimeSamplingRate: opts.EncodingTimeSamplingRate(),
+		encoder:                  protobuf.NewAggregatedEncoder(opts.BytesPool()),
+		p:                        producer,
+		numShards:                producer.NumShards(),
+		closed:                   false,
+		rand:                     rand.New(rand.NewSource(nowFn().UnixNano())),
+		metrics:                  newProtobufWriterMetrics(instrumentOpts.MetricsScope()),
+		nowFn:                    nowFn,
+	}
+	w.randFn = w.rand.Float64
+	w.shardFn = w.shard
+	return w, nil
+}
+
+func (w *protobufWriter) Write(mp aggregated.ChunkedMetricWithStoragePolicy) error {
+	if w.closed {
+		w.metrics.writerClosed.Inc(1)
+		return errWriterClosed
+	}
+	var encodeNanos int64
+	if w.encodingTimeSamplingRate > 0 && w.randFn() < w.encodingTimeSamplingRate {
```

Comment:
nit: Is the `w.encodingTimeSamplingRate > 0` check redundant since if `w.encodingTimeSamplingRate` is 0 then `w.randFn() < w.encodingTimeSamplingRate` will always be false?

Reply:
I moved this code from another place, I think it's still useful as when w.encodingTimeSamplingRate is less than 0 we skip the randFn() call